File: uva/bst.cpp
#include <stdio.h>
#include <vector>
using namespace std;

typedef struct Node {
  int val;
  struct Node* left;
  struct Node* right; 
  struct Node* parent;
  Node(int num, struct Node* leftChild, struct Node* rightChild, struct Node* nodeParent) {
    val = num;
    left = leftChild;
    right = rightChild;
    parent = nodeParent;
  }
} Node;

void postorder(Node* root) {
  vector<Node*> m;
  if (!root) {
    return;
  }
  Node* cur = root;
  while (1) {
    while (cur) {
      if (cur->right) {
        m.push_back(cur->right);
      }
      m.push_back(cur);
      cur = cur->left;
    } 
    cur = m.back();
    m.pop_back();
    if (!m.empty() && m.back() == cur->right) {
      m.pop_back();
      m.push_back(cur);
      cur = cur->right;
    }
    else {
      printf("%d\n", cur->val);
      cur = 0;
    }
    if (m.empty()) {
      break;
    }
  }
}

void inorder(Node* root) {
  if (root->left) {
    inorder(root->left);
  }
  printf("%d\n", root->val);
  if (root->right) {
    inorder(root->right);
  }
}

int main() {
  vector<Node*> m;
  int a;
  if (scanf("%d", &a) != 1) {
    return 0;
  }
  Node* root = new Node(a, 0, 0, 0);
  m.push_back(root);
  Node* node;
  while(scanf("%d", &a) == 1) {
    node = new Node(a, 0, 0, 0);
    if (m.back()->val >= node->val) {
      m.back()->left = node;
      node->parent = m.back();
      m.push_back(node);
      continue;
    }
    Node* top;
    while (m.back()->val < node->val) {
      top = m.back();
      m.pop_back();
      if (m.empty()) {
        break;
      }
    }
    top->right = node;
    node->parent = top;
    m.push_back(node);
  }
  postorder(root);
  return 0;
}

File: uva/bst.py
class Node:
    def __init__(self, val, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right

def post_order(root):
    if root == None:
        return
    post_order(root.left)
    post_order(root.right)
    print(root.val)

from sys import stdin
input = open("in.txt", "r")
preorder = []
while True:
    try:
        a = int(input.readline())
        preorder.append(a)
    except Exception as e:
        break
if preorder:
    root = Node(preorder[0])
    cur_stack = [root]
    cur_index = 0
    while cur_index < len(preorder) - 1:
        cur_index += 1
        p = cur_stack[-1]
        new_node = Node(preorder[cur_index])
        if new_node.val < p.val:
            p.left = new_node
            cur_stack.append(new_node)
            continue
        while cur_stack and new_node.val > cur_stack[-1].val:
            p = cur_stack[-1]
            cur_stack.pop()
        p.right = new_node
        cur_stack.append(new_node)
    #post_order(root)
    cur = root
    traverse_stack = []
    while True:
        while cur:
            if cur.right:
                traverse_stack.append(cur.right)
            traverse_stack.append(cur) 
            cur = cur.left
        cur = traverse_stack.pop() 
        if cur.right and traverse_stack and (traverse_stack[-1] == cur.right):
            traverse_stack.pop()
            traverse_stack.append(cur)
            cur = cur.right
        else:
            print(cur.val)
            cur = None
        if not traverse_stack:
            break
